File: src/modifiers/alert.ts
import type { ReactNode } from 'react';
import { createModifier } from './createModifier';

export interface AlertParams {
  /// The alert's title text.
  title: string;
  /// Whether the alert is shown. JS owns the source of truth; the native
  /// side mirrors it in local `@State` and converges on the JS value.
  isPresented: boolean;
  /// Fired with the new boolean whenever the presentation state changes
  /// (e.g. an action button or the system dismisses the alert).
  onChange?: (isPresented: boolean) => void;
  /// The alert's action buttons (typically `<Button>`s).
  actions: ReactNode;
  /// Optional descriptive message shown below the title.
  message?: ReactNode;
}

/// SwiftUI `.alert(_:isPresented:actions:message:)`. Presents an alert
/// titled `title` when `isPresented` is true. `actions` supplies the
/// buttons and `message` the optional body. The boolean is bound two-way
/// so a button tap / dismissal reports back to JS via `onChange`.
export function alert(params: AlertParams) {
  const { title, isPresented, onChange, actions, message } = params;
  return createModifier('alert', {
    title,
    isPresented,
    handler: onChange,
    actions,
    message,
  });
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/Alert.swift
import SwiftUI

/// SwiftUI `.alert(_:isPresented:actions:message:)`. Presents an alert
/// titled `title` when `isPresented` is true; the `actions` slot supplies
/// the buttons and the optional `message` slot the descriptive body.
/// This alert API is available on watchOS 9.0+ (the deployment target).
///
/// The presentation flag is bound bidirectionally with the same optimistic
/// pattern as `ToggleView`: a local `@State` mirror flips instantly when an
/// action button or the system dismisses the alert and fires the handler
/// with the new value, then converges to the next JS snapshot via
/// `.onChange(of:)`.
enum RNWAlertModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("alert") { view, params, ctx in
            let actions = ctx.content(params.string("actions"))
            let message = ctx.content(params.string("message"))
            return AnyView(view.modifier(RNWAlertViewModifier(
                title: params.string("title") ?? "",
                remotePresented: params.bool("isPresented") ?? false,
                handlerId: params.int("handler"),
                bus: ctx.bus,
                actionsBody: actions,
                messageBody: message
            )))
        }
    }
}

private struct RNWAlertViewModifier: ViewModifier {
    let title: String
    let remotePresented: Bool
    let handlerId: Int?
    let bus: RNWEventBus
    let actionsBody: AnyView?
    let messageBody: AnyView?

    @State private var localPresented: Bool

    init(
        title: String,
        remotePresented: Bool,
        handlerId: Int?,
        bus: RNWEventBus,
        actionsBody: AnyView?,
        messageBody: AnyView?
    ) {
        self.title = title
        self.remotePresented = remotePresented
        self.handlerId = handlerId
        self.bus = bus
        self.actionsBody = actionsBody
        self.messageBody = messageBody
        _localPresented = State(initialValue: remotePresented)
    }

    func body(content: Content) -> some View {
        let binding = Binding<Bool>(
            get: { localPresented },
            set: { newValue in
                localPresented = newValue
                if let handlerId {
                    bus.fire(handlerId, payload: newValue)
                }
            }
        )

        return content
            .alert(title, isPresented: binding) {
                actionsBody ?? AnyView(EmptyView())
            } message: {
                messageBody ?? AnyView(EmptyView())
            }
            .onChange(of: remotePresented) { newRemote in
                if newRemote != localPresented {
                    localPresented = newRemote
                }
            }
    }
}
